rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,207 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+public class RemoteFetchProcessor extends AbstractScheduledService
+{
+  private Configuration conf;
+  BookKeeper bookKeeper;
+  private Queue<FetchRequest> processQueue = null;
+  private ConcurrentMap<String, RangeSet<Long>> rangeMap = null;
+  private ConcurrentMap<String, FileMetadataRequest> fetchRequestMap = null;
+  private Runnable runnableStatusUpdateTask = null;
+  private ExecutorService processService;
+
+  private static DirectBufferPool bufferPool = new DirectBufferPool();
+
+  int diskReadBufferSize;
+  int numRemoteFetchThreads;
+  int remoteFecthThreadInitialDelay;
+  int remoteFetchThreadInterval;
+  int processThreadInitalDelay;
+  int processThreadInterval;
+  long requestProcessDelay;
+
+  private static final Log log = LogFactory.getLog(RemoteFetchProcessor.class);
+
+  public RemoteFetchProcessor(BookKeeper bookKeeper, Configuration conf)
+  {
+    this.bookKeeper = bookKeeper;
+    this.conf = conf;
+    this.processQueue = new ConcurrentLinkedQueue<FetchRequest>();
+    this.rangeMap = new ConcurrentHashMap<String, RangeSet<Long>>();
+    this.fetchRequestMap = new ConcurrentHashMap<String, FileMetadataRequest>();
+
+    this.diskReadBufferSize = CacheConfig.getDiskReadBufferSizeDefault(conf);
+    this.processThreadInitalDelay = CacheConfig.getProcessThreadInitialDelayInMs(conf);
+    this.processThreadInterval = CacheConfig.getProcessThreadIntervalInMs(conf);
+    this.requestProcessDelay = CacheConfig.getRemoteFetchProcessIntervalInMS(conf);
+    int numThreads = CacheConfig.getRemoteFetchNumThreads(conf);
+
+    ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(numThreads);
+    processService = MoreExecutors.getExitingExecutorService(executor);
+  }
+
+  public void addToProcessQueue(String remotePath, long offset, int length, long fileSize, long lastModified)
+  {
+    long requestedTime = System.currentTimeMillis();
+    FetchRequest request = new FetchRequest(remotePath, offset, length, fileSize, lastModified, requestedTime);
+    processQueue.add(request);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception
+  {
+    long currentTime = System.currentTimeMillis();
+
+    // Till the queue is not empty or there are no more requests which came in before the configured delay time
+    // we are going to collect the requests and process them
+    while (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.peek();
+      if (currentTime - request.getRequestedTime() < this.requestProcessDelay) {
+        break;
+      }
+
+      if (!rangeMap.containsKey(request.getRemotePath())) {
+        RangeSet<Long> rangeSet = TreeRangeSet.create();
```

Comment:
Is this the code snippet that eventually combines ranges/batches reads ? Can you add unit tests for these two ? What will help is a list of requests and tests that show how they look after they are batched. This will also force you to move this code - while loop - to a separate function.